### ISSUE TITLE
Add Utils\Helper::isValid() for easier validation

### ISSUE THREAD

**Issue by Art4** [Art4/json-api-client#15]
```php
$wrong_jsonapi = '{"data":{},"meta":{"info":"This is wrong JSON API. `data` has to be `null` or containing at least `type` and `id`."}}';

if ( ! \Art4\JsonApiClient\Utils\Helper::isValid($wrong_jsonapi) )
{
    // string is not in valid jsonapi format.
}
```

**Comment by fcastillo-es** [comment]
I like the idea. Will this be able to validate fragments?

It would be great for me if I could call isValid for any resource, so if I override a class to be editable, I could extend and call the validation after I finish the modifications, or when I get the document as array or as string instead of in the constructor.

**Comment by Art4** [comment]
> Will this be able to validate fragments?

Interesting idea. I thought only about a whole jsonapi string but it must be possible to validate only fragments too. We could add an optional second parameter to define the entry point.

Example:

```php
if ( ! \Art4\JsonApiClient\Utils\Helper::isValid($json_string, 'Resource\Item') )
{
    // $json_string is not a valid Resource\Item
}
```

On the other hand it would be difficult to validate a `RelationshipCollection` because it requires a Resource in the constructor at the moment. This must be changed first and need some more improvements.

>  I could extend and call the validation after I finish the modifications, or when I get the document as array or as string instead of in the constructor.

I would prefer only validate a json string. A pure array without objects isn't valid json api in most cases.

**Comment by fcastillo-es** [comment]
>Interesting idea. I thought only about a whole jsonapi string but it must be possible to validate only fragments too. We could add an optional second parameter to define the entry point.

Sounds good to me.

>I would prefer only validate a json string. A pure array without objects isn't valid json api in most cases.

The idea would not be to validate a pure array, I was thinking about validating the types of the properties in the Document recursively before returning the array. It would be similar to what you have in Document::parseData, where you get the structure from json_decode and use it to assign Collection, Item or throw ValidationException as needed, but the other way.

Next week I'm on vacation, but I'll think about something like each class defining its validation rules and implementing a ValidationTrait.

```php
// valid types
$validations = array(
    'data' => array('Collection', 'Identifier', 'NullResource'),
    // (...)
);
```

This way I could, for example, use a fluent interface to build an item and validate it when trying to assign it:

```php
$data = new Item()
$data->setId('123')
    ->setType('user')
    ->addAttribute('email', 'test@example.com');
$document->setData($data); // trigger validation and assign
```

I'll most surely need it for my project, but I don't know if it wold fit yours. Tell me what you think about it to know if I keep it or share something.

**Comment by Art4** [comment]
Sorry for the late response. I was working at another project.

This would be out of scope of this project. This library should be a pure client to handle/validate the JSON API response. So I'm not planning to include setters or other functionality to build an object, sorry This should be done through the factory.

But feel free to open issues if you need more possibilities to extend the library.

**Comment by fcastillo-es** [comment]
Don't worry, I was also working in another project.

Ok, then I'll create another repository with the classes and functionality I need and give you feedback about the extension feature.

Thanks.

**Comment by fcastillo-es** [comment]
Hi, I've finally decided to go with a modified version of your client, mostly because adding fractal to get the output would add a second validation, which I would like to avoid.

I'm doing some work extracting validators outside of the main classes. First just to create concrete Validator classes, then maybe add a Validator class similar to the Factory, with the list of classes for each resource. That way the manager could also manage the validator, and that way the code could support switching validations. For example, a client could use the request headers to load validators for a specific version of json-api.

This is the first commit in that line: https://github.com/fcastillo-es/json-api-client/commit/279fb0fb92e249c371a536584ede06432090d01b. As I think it's related with this issue please tell me if you would accept this kind of changes if I complete them and add unit testing. Else I would just carry on with my fork.

Thanks.

**Comment by Art4** [comment]
I'm not sure if it would be as easy to put all validation in their own classes, but I'm interested to see more.

**Comment by fcastillo-es** [comment]
I am working slowly due to other project, but it is all in a branch in my fork https://github.com/fcastillo-es/json-api-client/tree/external-validator.

The first step is Creating the validator classes. After that, I want to add a Validator class to add to the manager and use the same strategy of the class mapping so the actual validator may be switched. That way the object will only need to take the manager in the constructor and they can get validation from there.

**Comment by fcastillo-es** [comment]
Hi, sorry for having said nothing before, but we're moving away from json api, so I will not be able to carry on working on the client for now. Thanks for all.

**Comment by Art4** [comment]
No problem, thanks for your help. :+1:  I'm glad I could help you.

**Comment by Art4** [comment]
Was implemented with cfa7f1d in release 0.6
